### PR TITLE
Add compatibility test to ensure we don't break VS.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Manually took a snapshot of the current `TagHelperDescriptor` for Razor and prefixed each bit with `VisualStudio`.
- Created a test that took the current runtime `TagHelperDescriptor` serialized it then attempts to deserialize it back into the `VisualStudio` version.

#12

### CODE REVIEW

**Comment by Eilon** [comment]
Seems pretty straightforward, :shipit:

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+            Assert.Equal(attribute.PropertyName, expectedAttribute.PropertyName, StringComparer.Ordinal);
+            Assert.Equal(attribute.TypeName, expectedAttribute.TypeName, StringComparer.Ordinal);
+
+            var dtAttribute = attribute.DesignTimeDescriptor;
+            var expectedDTAttribute = expectedAttribute.DesignTimeDescriptor;
+            Assert.Equal(dtAttribute.Remarks, expectedDTAttribute.Remarks, StringComparer.Ordinal);
+            Assert.Equal(dtAttribute.Summary, expectedDTAttribute.Summary, StringComparer.Ordinal);
+        }
+
+        #region PinnedVisualStudioTagHelperDescriptors
```
Comment (dougbu) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Why a `#region`?

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+                get
+                {
+                    return _name;
+                }
+                set
+                {
+                    if (value == null)
+                    {
+                        throw new ArgumentNullException(nameof(value));
+                    }
```
Comment (dougbu) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Why `null` check anything here? Suggest making everything auto-properties.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+                    _propertyName = value;
+                }
+            }
+            public string TypeName
```
Comment (dougbu) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Bit 'o blank lines betwixt the members.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+                    {
+                        throw new ArgumentNullException(nameof(value));
+                    }
+
+                    _typeName = value;
+                    IsStringProperty = string.Equals(TypeName, typeof(string).FullName, StringComparison.Ordinal);
+                }
+            }
+            public VisualStudioTagHelperAttributeDesignTimeDescriptor DesignTimeDescriptor { get; set; }
+            public bool IsNameMatch(string name)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Don't need this method.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+            public IEnumerable<string> AllowedChildren { get; set; }
+            public string RequiredParent { get; set; }
+            public VisualStudioTagStructure TagStructure { get; set; }
+            public VisualStudioTagHelperDesignTimeDescriptor DesignTimeDescriptor { get; set; }
+        }
+        public enum VisualStudioTagStructure
+        {
+            Unspecified,
+            NormalOrSelfClosing,
+            WithoutEndTag
```
Comment (dougbu) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Corner cases but the test only confirms the `int` value for `TagStructure.NormalOrSelfClosing` remains `1`. It won't catch new `TagStructure` values, changing `Unspecified` to equal `-1` or changing `WithoutEndTag` to equal `37`.

**Comment by NTaylorMullen** [comment]
@dougbu the snapshot is a direct copy/paste from the current version of Razor minus the doc comments. I don't intend to modify it further.

**Comment by dougbu** [comment]
> I don't intend to modify it further.

I don't get this. Let's talk offline.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs (hunk 1 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Because that code "isn't really here" would be my guess.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs (hunk 2 above)
Comment (Eilon) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: I don't think this code matters, it just mirrors the runtime, no?

**Comment by Eilon** [comment]
If I understood correctly, this test just exists to ensure we don't break ourselves. As such, it would be counterproductive to change the "VisualStudioXYZ" implementations because then it would no longer mirror the current runtime versions of "XYZ".

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs (hunk 1 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Yup.

File: test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs (hunk 2 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Tooling.Razor.Test/VisualStudioProtocolTest.cs]: Yup

**Comment by NTaylorMullen** [comment]
@Eilon you understood correctly :smile:
